### PR TITLE
Add `default_enabled` option to optional message dict

### PR DESCRIPTION
**Pull request by cdce8p**
## Description
Add a new `default_enabled` option to make it easier to disable a specific message by default.
This is especially useful for extensions like the `CodeStyle` one. I'm not so sure if we should use it for regular checks though. The expectation I believe is that all of them are enabled by default. I.e. to disable one, we should move it to an extension first.

### CODE REVIEW

**Comment by DanielNoord** [comment]
This feels like it fixes part of https://github.com/PyCQA/pylint/issues/3512 but not fully.

I'm -0,5 for adding any more complexity to enabling and disabling that doesn't immediately solve that issue. It just creates more to consider before we implement the system discussed there.

File: pylint/lint/pylinter.py
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:485]: It feel like it's the responsibility of the Message Store to not register disabled messages in ``register_messages_from_checker``.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I think it's a good idea, we definitely need an easy way to make a check optional (beside moving code to another directory and adding or removing code during registration, that's barbaric 😄). We need a better way but plugin maintainer also need a better way imo.

I would use a positive formulation for the option name to avoid a double negative (disabled = False => enabled = True), maybe ``default_enabled`` or ``enabled_by_default`` ?

File: pylint/lint/pylinter.py
Comment (cdce8p) [reply pylint/lint/pylinter.py]: We want to register all messages. Otherwise the user isn't able to enable it via the config setting. It's the `MessageStateHandler` which deals with `enable`, `disable`, and `disable-next`.

**Review by cdce8p — COMMENTED** [review]
> I think it's a good idea, we definitely need an easy way to make a check optional (beside moving code to another directory and adding or removing code during registration, that's barbaric 😄). We need a better way but plugin maintainer also need a better way imo.

I think we should be careful when considering if a check should be disabled by default. It's much easier for new ones and extensions. However, I do believe there is at least the expectation that all default checks are also enabled.

That doesn't negate that having an easy option to do so would be useful IMO.

> I would use a positive formulation for the option name to avoid a double negative (disabled = False => enabled = True), maybe `default_enabled` or `enabled_by_default` ?

Done

**Comment by Pierre-Sassoulas** [comment]
> I think we should be careful when considering if a check should be disabled by default.

Sure, definitely something that will help for #3512 but should not change much after that.

File: pylint/lint/pylinter.py
Comment (cdce8p) [reply pylint/lint/pylinter.py]: > Hmm, right. It feels wrong that this code is in `PyLinter`,

I don't think so. A few lines below we also disable messages if the complete checker is disabled.
https://github.com/PyCQA/pylint/blob/77e8ae4f3902697babb186482f577a2c8b1a3725/pylint/lint/pylinter.py#L483-L485

> ... it should probably happen somewhere in `pylint.message`. Do you think moving that to MessageStateHandler is possible ?

Not without some bigger refactoring. `MessageStateHandler` is part of the `pylint.lint` package and not related to `pylint.messages`. That's why `PyLinter` is the best place IMO. It's the connection between both.

Comment (Pierre-Sassoulas) [reply pylint/lint/pylinter.py]: > PyLinter is the best place IMO. It's the connection between both.

It's the connection between everything and is still way too huge even after Daniel most welcome refactors to separate the configuration parsing. But yeah, if it's a huge refactor then let's not do it right now.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by DanielNoord** [comment]
But don't want we want to plan out that PR and then work based on that design? We're not sure how this would interact with the extending/disabling that is proposed in that issue.

**Comment by Pierre-Sassoulas** [comment]
I'm pretty sure a boolean flag on each message is the simplest / most extensible / scalable way to disable / enable messages. Actually I've wanted to do that for a long time, (thank you @cdce8p). But I'm open to another design if you have one Daniel

**Comment by Pierre-Sassoulas** [comment]
Could you explain what you meant and the advantages of the template approach @DanielNoord ? I'm not sure about what it mean.

**Comment by DanielNoord** [comment]
Well just the whole discussion in that issue:
Do we want to create an `extend-disable` option?
Do we want to create shareable configuration templates? If so, how should these two interact.

The implementation of those questions might be more complicated if we add more stuff right now.

**Comment by Pierre-Sassoulas** [comment]
Ha thank you I get it know. I think this MR won't impact the shareable template configuration, this is only a way for pylint to internally change our default values easily. I think we should not expose it in configuration templates.

**Comment by cdce8p** [comment]
> Ha thank you I get it know. I think this MR won't impact the shareable template configuration, this is only a way for pylint to internally change our default values easily. I think we should not expose the `default_enabled` option in configuration templates.

I agree. The way I think about it, there are / will be multiple layers:
1. The default configuration (`default_enabled` being part of it)
2. Template / profiles which can overwrite the default configuration in a predefined manner. Similar to [isort profiles](https://pycqa.github.io/isort/docs/configuration/profiles.html).
3. Custom overwrites by the user.

Since `default_enabled` would be part of the default config, it shouldn't impact profiles at all. Similar to the way profiles won't change message ids or descriptions.

**Comment by DanielNoord** [comment]
Yeah, but it might interact with how we implement `extend-enable` which in turn might depend on templating. If you both think this is fine please go ahead and merge, but I just wanted to point this might have unforeseen consequences for future implementations,

**Comment by Pierre-Sassoulas** [comment]
I think the separation of default configuration / templated configuration /  user overwrite will prevent most issue with this, I'll let you merge if you agree @cdce8p.

**Comment by cdce8p** [comment]
> I think the separation of default configuration / templated configuration / user overwrite will prevent most issue with this, I'll let you merge if you agree @cdce8p.

I agree. It's good to be careful but I don't see an issue with `default_enabled` going forward. It just represents the lowest config level, i.e. the default configuration. Anything else will be on top of that.

**Pull request merged**
